Support reading ECP style policy config from a git url

simonbaird opened the pull request:
JIRA: [HACBS-2178](https://issues.redhat.com/browse/HACBS-2178)

joejstuart reviewed (approved):
Very neat. Maybe just some unit tests in `image_test.go`?

simonbaird commented:
> some unit tests in `image_test.go`?

I can't figure out a good way to do it, but will have another look.

zregvart commented on acceptance/git/git.go:
```diff
@@ -240,6 +241,12 @@ func createGitRepository(ctx context.Context, repositoryName string, files *godo
+		// Replace ${GITHOST} with the real git host.
+		// Used in acceptance/examples/default-policy.yaml
+		// Hopefully a noop for all other files.
+		re := regexp.MustCompile(regexp.QuoteMeta("${GITHOST}"))
+		b = []byte(re.ReplaceAllString(string(b), Host(ctx)))
```
Could be done using [os.Expand](https://pkg.go.dev/os#Expand) as we do in [other parts](https://github.com/enterprise-contract/ec-cli/blob/fdfc2f2890c7adfac3d17d7b01c001234aa131b4/acceptance/cli/cli.go#L144-L146) of acceptance tests

simonbaird replied:
Nice, thanks.

simonbaird replied:
Done. (In incoming revision).

zregvart commented on acceptance/git/git.go:
```diff
@@ -240,6 +241,12 @@ func createGitRepository(ctx context.Context, repositoryName string, files *godo
+		// Replace ${GITHOST} with the real git host.
+		// Used in acceptance/examples/default-policy.yaml
+		// Hopefully a noop for all other files.
+		re := regexp.MustCompile(regexp.QuoteMeta("${GITHOST}"))
+		b = []byte(re.ReplaceAllString(string(b), Host(ctx)))
```
Add a test for this, I'm fairly certain that `Host()` will return `""` at this point, i.e. before [line 177](https://github.com/simonbaird/ec-cli/blob/48fb0aadf463ab2ed95697e0db4528a0ea44f336/acceptance/git/git.go#L177)

simonbaird replied:
From what I can tell, it's fine since `startStubGitServer` via "Given stub git daemon running" calls that code, and that happens early on.

zregvart commented on cmd/validate/image.go:
```diff
@@ -127,6 +131,12 @@ func validateImageCmd(validate imageValidationFunc) *cobra.Command {
 
+			Use a git url for the policy configuration. In this example there should be a file called
+			'policy.yaml' inside a directory called 'default' in the 'github.com/acme-company/ec-configs'
+			git repo. For git repos that are not on GitHub, prefix the url with 'git::'
```
Or [gitlab.com](https://github.com/hashicorp/go-getter/blob/main/detect_gitlab.go)

simonbaird replied:
Done. (In incoming revision).

zregvart commented on cmd/validate/image.go:
```diff
@@ -170,6 +180,58 @@ func validateImageCmd(validate imageValidationFunc) *cobra.Command {
 			}
 
+			// Check if policyConfiguration is a git url, if so then we need to download the config
+			if strings.HasPrefix(data.policyConfiguration, "git::") || strings.HasPrefix(data.policyConfiguration, "github.com/") || strings.HasPrefix(data.policyConfiguration, "https://github.com/") {
```
Since we're adding more logic here this and the branch below should be responsibility of `github.com/enterprise-contract/ec-cli/internal/policy/source`

simonbaird replied:
Done. (In incoming revision).

zregvart commented on configs/default/policy.yaml:
```diff
@@ -0,0 +1,27 @@
+#
+description: >
+  Use the policy rules from the "minimal" collection. This and other collections are defined in
+  https://redhat-appstudio.github.io/docs.stonesoup.io/ec-policies/release_policy.html#_available_rule_collections
```
Should this URL and the one below go to enterprisecontract.dev?

simonbaird replied:
Done.

zregvart commented on configs/default/policy.yaml:
```diff
@@ -0,0 +1,27 @@
+publicKey: "k8s://tekton-chains/public-key"
+
+sources:
+  - name: Default RHTAP Release Policies
```
Should this say Enterprise Contract instead of RHTAP?

simonbaird replied:
I made it just "Default".

simonbaird commented:
Thanks @zregvart for the multiple good suggestions. New revision summary:
* I picked out some small unrelated refactors into their own commits, and also pulled in https://github.com/simonbaird/ec-cli/pull/1 .
* Move the git config downloading and some related logic in to `internal/policy/source/git_config.go`
* Added some unit tests
* Added some more acceptance tests, including one for an error condition
* It now looks in more than one place for the config file, either `.ec/policy.ext` or `policy.ext` where ext can be one of json, yaml or yml.

simonbaird commented on internal/policy/source/source.go:
```diff
@@ -37,6 +42,7 @@ const (
 	DownloaderFuncKey key        = 0
 	PolicyKind        policyKind = "policy"
 	DataKind          policyKind = "data"
+	ConfigKind        policyKind = "config"
```
Actually we could use either one of the other kinds and would work the same, but I wanted to add this for clarity of intent.

zregvart reviewed (approved):
LGTM
